code/modules/projectiles/guns/beam_rifle.dm

```diff
@@ -0,0 +1,418 @@
+	set waitfor = 0
+	if(!log_override && firer && original)
+		add_logs(firer, original, "fired at", src, " [get_area(src)]")
+	if(setAngle)
+		Angle = setAngle
+	var/next_run = world.time
+	var/old_pixel_x = pixel_x
+	var/old_pixel_y = pixel_y
+	while(loc)
+		if((!( current ) || loc == current))
+			current = locate(Clamp(x+xo,1,world.maxx),Clamp(y+yo,1,world.maxy),z)
+		if(!Angle)
+			Angle=round(Get_Angle(src,current))
+		if(spread)
+			Angle += (rand() - 0.5) * spread
+		var/matrix/M = new
+		M.Turn(Angle)
+		transform = M
+		var/Pixel_x=sin(Angle)+16*sin(Angle)*2
+		var/Pixel_y=cos(Angle)+16*cos(Angle)*2
+		var/pixel_x_offset = old_pixel_x + Pixel_x
+		var/pixel_y_offset = old_pixel_y + Pixel_y
+		var/new_x = x
+		var/new_y = y
+		while(pixel_x_offset > 16)
+			pixel_x_offset -= 32
+			old_pixel_x -= 32
+			new_x++// x++
+		while(pixel_x_offset < -16)
+			pixel_x_offset += 32
+			old_pixel_x += 32
+			new_x--
+		while(pixel_y_offset > 16)
+			pixel_y_offset -= 32
+			old_pixel_y -= 32
+			new_y++
+		while(pixel_y_offset < -16)
+			pixel_y_offset += 32
+			old_pixel_y += 32
+			new_y--
+		pixel_x = old_pixel_x
+		pixel_y = old_pixel_y
+		step_towards(src, locate(new_x, new_y, z))
+		next_run += max(world.tick_lag, speed)
+		var/delay = next_run - world.time
+		if(delay <= world.tick_lag*2)
+			pixel_x = pixel_x_offset
+			pixel_y = pixel_y_offset
+		else
+			animate(src, pixel_x = pixel_x_offset, pixel_y = pixel_y_offset, time = max(1, (delay <= 3 ? delay - 1 : delay)), flags = ANIMATION_END_NOW)
+		old_pixel_x = pixel_x_offset
+		old_pixel_y = pixel_y_offset
+		if(original && (original.layer>=2.75) || ismob(original))
+			if(loc == get_turf(original))
+				if(!(original in permutated))
+					Bump(original, 1)
+		Range()
+
+/obj/item/projectile/beam/beam_rifle/hitscan/Range()
+	spawn_tracer_effect()
+
+/obj/item/projectile/beam/beam_rifle/hitscan/proc/spawn_tracer_effect()
+	new tracer_type(loc, time = 5, angle_override = Angle, p_x = pixel_x, p_y = pixel_y, color_override = color)
+
+/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam
+	tracer_type = /obj/effect/temp_visual/projectile_beam/tracer/aiming
+	name = "aiming beam"
+	hitsound = null
+	hitsound_wall = null
+
+/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/prehit()
+	qdel(src)
+	return FALSE
+
+/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/spawn_tracer_effect()
+	var/obj/effect/temp_visual/projectile_beam/T = new tracer_type(loc, time = 5, angle_override = Angle, p_x = pixel_x, p_y = pixel_y, color_override = color)
+	if(istype(gun) && istype(T))
+		gun.current_tracers[T] = TRUE
+
+/obj/effect/temp_visual/projectile_beam
+	icon = 'icons/obj/projectiles.dmi'
+	layer = ABOVE_MOB_LAYER
+	anchored = 1
+	duration = 5
+	randomdir = FALSE
+
+/obj/effect/temp_visual/projectile_beam/New(time = 5, angle_override, p_x, p_y, color_override)
+	duration = time
+	if(isnull(angle_override)||isnull(p_x)|isnull(p_y))
+		qdel(src)
+	pixel_x = p_x
```

Comment:
The difference is minimal so i'd say 1+ just so it's easier to notice the pattern for new people.